server/games/game.py

```diff
@@ -540,39 +542,76 @@ def clear_slot(self, slot_index):
-            await self.mark_invalid(ValidityState.WRONG_VICTORY_CONDITION)
+        # Only allow ranked mods
+        for mod_id in self.mods.keys():
+            if mod_id not in self.game_service.ranked_mods:
+                await self.mark_invalid(ValidityState.BAD_MOD)
+                return
 
-        if self.has_ai or (self.gameOptions["AIReplacement"] != "Off"):
+        if self.has_ai:
             await self.mark_invalid(ValidityState.HAS_AI_PLAYERS)
-        elif self.is_multi_team:
+            return
+        if self.is_multi_team:
             await self.mark_invalid(ValidityState.MULTI_TEAM)
-        elif self.is_ffa:
+            return
+        if self.is_ffa:
             await self.mark_invalid(ValidityState.FFA_NOT_RANKED)
-        elif not self.is_even:
-            await self.mark_invalid(ValidityState.UNEVEN_TEAMS_NOT_RANKED)
-        elif self.gameOptions["FogOfWar"] != "explored":
-            await self.mark_invalid(ValidityState.NO_FOG_OF_WAR)
-
-        elif self.gameOptions["CheatsEnabled"] != "false":
-            await self.mark_invalid(ValidityState.CHEATS_ENABLED)
+            return
+        valid_options = {
+            "AIReplacement": ("Off", ValidityState.HAS_AI_PLAYERS),
+            "FogOfWar": ("explored", ValidityState.NO_FOG_OF_WAR),
+            "CheatsEnabled": ("false", ValidityState.CHEATS_ENABLED),
+            "PrebuiltUnits": ("Off", ValidityState.PREBUILT_ENABLED),
+            "NoRushOption": ("Off", ValidityState.NORUSH_ENABLED),
+            "RestrictedCategories": (0, ValidityState.BAD_UNIT_RESTRICTIONS),
+            "TeamLock": ("locked", ValidityState.UNLOCKED_TEAMS)
+        }
+        if await self._validate_game_options(valid_options) is False:
+            return
 
-        elif self.gameOptions["PrebuiltUnits"] != "Off":
-            await self.mark_invalid(ValidityState.PREBUILT_ENABLED)
+        if self.game_mode in ('faf', 'ladder1v1'):
+            await self._validate_faf_game_settings()
+        elif self.game_mode == 'coop':
+            await self._validate_coop_game_settings()
+
+    async def _validate_game_options(self,
+                                     valid_options: Dict[str, Tuple[Any, ValidityState]]) -> bool:
+        for key, value in self.gameOptions.items():
+            if key in valid_options:
+                (valid_value, validity_state) = valid_options[key]
+                if self.gameOptions[key] != valid_value:
+                    await self.mark_invalid(validity_state)
+                    return False
+        return True
 
-        elif self.gameOptions["NoRushOption"] != "Off":
-            await self.mark_invalid(ValidityState.NORUSH_ENABLED)
+    async def _validate_coop_game_settings(self):
+        """
+        Checks which only apply to the coop mode
+        """
+        # Note, the UNEVEN_TEAMS_NOT_RANKED state is used as a placeholder for reasons
```

Comment:
This sounds really dirty to me. I would prefer you to add all of those  conditions in the DB via a migration scripts, and to add them to server code later.